shadowsocks/dialer.go

```diff
@@ -0,0 +1,196 @@
+func NewDialer(host string, port int, password, cipher string) (Dialer, error) {
+	proxyIP, err := net.ResolveIPAddr("ip", host)
+	if err != nil {
+		return nil, errors.New("Failed to resolve proxy address")
+	}
+	aead, err := newAeadCipher(cipher, password)
+	if err != nil {
+		return nil, err
+	}
+	d := ssDialer{proxyIP: proxyIP.IP, proxyPort: port, cipher: aead}
+	return &d, nil
+}
+
+func (d *ssDialer) DialTCP(laddr *net.TCPAddr, raddr string) (onet.DuplexConn, error) {
+	socksTargetAddr := socks.ParseAddr(raddr)
+	if socksTargetAddr == nil {
+		return nil, errors.New("Failed to parse target address")
+	}
+	proxyAddr := &net.TCPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	proxyConn, err := net.DialTCP("tcp", laddr, proxyAddr)
+	if err != nil {
+		return nil, err
+	}
+	ssw := NewShadowsocksWriter(proxyConn, d.cipher)
+	_, err = ssw.Write(socksTargetAddr)
+	if err != nil {
+		proxyConn.Close()
+		return nil, errors.New("Failed to write target address")
+	}
+	ssr := NewShadowsocksReader(proxyConn, d.cipher)
+	return onet.WrapConn(proxyConn, ssr, ssw), nil
+}
+
+// Clients can use the io.ReadWriter methods of onet.PacketConn to leverage the connection to `raddr`.
+func (d *ssDialer) DialUDP(laddr *net.UDPAddr, raddr string) (onet.PacketConn, error) {
+	targetHost, targetPort, err := SplitHostPortNumber(raddr)
+	if err != nil {
+		return nil, err
+	}
+	targetAddr := &packetConnAddr{Host: targetHost, Port: targetPort}
+	proxyAddr := &net.UDPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	pc, err := net.ListenUDP("udp", laddr)
+	if err != nil {
+		return nil, err
+	}
+	conn := packetConn{
+		PacketConn: pc, proxyAddr: proxyAddr, targetAddr: targetAddr,
+		cipher: d.cipher, buf: make([]byte, udpBufSize)}
+	return &conn, nil
+}
+
+type packetConn struct {
+	net.PacketConn
+	io.ReadWriter
+	proxyAddr  *net.UDPAddr
+	targetAddr net.Addr
+	cipher     shadowaead.Cipher
+	m          sync.Mutex
+	buf        []byte // Write lock
+}
+
+// Write encrypts `b` and writes to the connected address through the proxy.
+func (c *packetConn) Write(b []byte) (int, error) {
+	return c.WriteTo(b, c.targetAddr)
+}
+
+// WriteTo encrypts `b` and writes to `addr` through the proxy.
+func (c *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	socksTargetAddr := socks.ParseAddr(addr.String())
+	if socksTargetAddr == nil {
+		return 0, errors.New("Failed to parse target address")
+	}
+	buf, err := shadowaead.Pack(c.buf, append(socksTargetAddr, b...), c.cipher)
+	if err != nil {
+		return 0, err
+	}
+	_, err = c.PacketConn.WriteTo(buf, c.proxyAddr)
+	return len(b), err
+}
+
+func (c *packetConn) Read(b []byte) (int, error) {
+	n, _, err := c.ReadFrom(b)
+	return n, err
+}
+
+// ReadFrom reads from the embedded PacketConn and decrypts into `b`.
+func (c *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
+	if len(b) < c.cipher.SaltSize() {
+		return 0, nil, shadowaead.ErrShortPacket
+	}
+	n, addr, err := c.PacketConn.ReadFrom(b)
```

Comment:
This seems slightly wrong.  `ReadFrom` will fail if `b` is too short to hold the encrypted packet, even if `b` is actually long enough to hold the decrypted contents.

I think it would be more correct to read into a buffer that is long enough to hold encrypted plaintext of length `cap(b)`.

Reply:
N/A, we now read into a MaxUDPBufferSize buffer.